### PR TITLE
A new strategy for generating functions

### LINKED ISSUE #167: Generating functions
One of the nice things Quickcheck can do is generate random functions. It would be nice if Hypothesis could do the same.

### PR DESCRIPTION
**Pull request by Zac-HD**
It's the simplest possible thing that closes #167.

`functions(like=lambda: None, returns=none())` takes two arguments; the first must be a callable and the second must be a strategy, both are optional.  The generated function will pretend to be `like`, and return a value drawn from `returns` each time it is called.  This strategy supports `.example()`.

There are multiple ways that this could be improved, such as simulating pure functions or inferring a return strategy from the return type hint of `like`, but nothing that I can think of would break this API.  I therefore decided to get a MVP out first, and leave the harder parts for after someone actually asks for them.

### CODE REVIEW

**Comment by DRMacIver** [comment]
> This strategy supports `.example()`.

I don't think it does. It will end up eventually throwing `StopTest`, leaking an internal implementation to the outside. It's exactly the sort of thing that *shouldn't* support `example` really.

File: hypothesis-python/src/hypothesis/_strategies.py [hunk 1]
```diff
@@ -2189,3 +2187,33 @@ def emails():
     return builds(u"{}@{}".format, local_part, domains()).filter(
         lambda addr: len(addr) <= 254
     )
+
+
+# Mypy can't yet handle default values with generic types or typevars, but the
+# @overload workaround from https://github.com/python/mypy/issues/3737 doesn't
+# with @composite functions - Mypy can't see that the function does in fact
```
Comment (alexwlchan) [inline hypothesis-python/src/hypothesis/_strategies.py]: nit: did you mean to say “doesn’t *work* with @composite functions”?

File: hypothesis-python/tests/cover/test_functions.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    with pytest.raises(InvalidState):
+        cache[0]()
+
+
+def test_can_call_default_like_arg():
+    # This test is somewhat silly, but coverage complains about the uncovered
+    # branch for calling it otherwise and alternative workarounds are worse.
+    like, returns = getfullargspec(functions).defaults
+    assert like() is None
+    assert returns.example() is None
```
Comment (alexwlchan) [inline hypothesis-python/tests/cover/test_functions.py]: What if a define a function like so (in Python 3-only syntax, granted):

```python
def func_d(a, *, b, c):
    pass
```

Will the function that Hypothesis generates:

* Allow passing argument `a` positionally or as a keyword
* Only allow passing `b` or `c` as kwargs

Would be nice to have a (Python 3-only) test for this.

Comment (Zac-HD) [reply hypothesis-python/tests/cover/test_functions.py]: Done :smile: 

Thanks for the review!

**Comment by Zac-HD** [comment]
CI failure is due to `pip` 19.1 breaking the current version of `pip-compile`; I'm just going to wait for jazzband/pip-tools#795 instead of trying a workaround.

**Pull request merged**